File: 09-07/euler21.py
def euler21():
	
	# Returns the sum of the divisors of n 
	def sumDivs(n):
		return sum([i for i in range(1, n) if n % i == 0])

	# Returns a tuple containing the sumDivs(n) and n if sumDivs(sumDivs(n)) is equal to n, otherwise (None, None)
	def amicable(n):
		total = sumDivs(n)
		return (total, n) if sumDivs(total) == n else (None, None)

	# Initialize seen list, keeps track of all amicable numbers we have seen in the past
	seen = []
	total = 0

	# Loop through all numbers under 10,000
	for i in range(10000):
		x,y = amicable(i)
		# Problem specifies that x cannot equal y, also if x is None, then the number is not amicable
		if x == y or x == None: continue
		# Do not add repeats to the total
		if x not in seen:
			# Add numbers to total and to list of seen numbers
			total += x + y
			seen.extend([x,y])
	return total

File: 09-07/euler21/euler21.cpp
#include <iostream>
#include <vector>

int sumDivs(int n);
bool contains(std::vector<int> v, int n);
int amicable(int n);
int euler21(void);

int main(void) {
	std::cout << euler21() << std::endl;
	return 0;
}

int sumDivs(int n) {
	int sum = 0;
	for (int i = 1; i < n; i++) {
		if (n % i == 0) {
			sum += i;
		}
	}
	return sum;
}

bool contains(std::vector<int> v, int n) {
	for (int i = 0; i < v.size(); i++) {
		if (v[i] == n) {
			return true;
		}
	}
	return false;
}

int amicable(int n) {
	int total = sumDivs(n);
	if (sumDivs(total) == n) {
		return total;
	}
	return -1;
}

int euler21(void) {

	std::vector<int> seen;
	int total = 0;

	for (int i = 0; i < 10000; i++) {

		int x = amicable(i);

		if (x != -1 && x != i) {
			
			if (!contains(seen, x)) {
				total += x + i;
				seen.push_back(x);
				seen.push_back(i);
			}

		}

	}
	return total;
}
